components/camel-telegram/src/main/java/org/apache/camel/component/telegram/TelegramConsumer.java

```diff
@@ -0,0 +1,109 @@
+import org.apache.camel.Processor;
+import org.apache.camel.component.telegram.model.Update;
+import org.apache.camel.component.telegram.model.UpdateResult;
+import org.apache.camel.impl.ScheduledPollConsumer;
+
+/**
+ * A polling consumer that reads messages from a chat using the Telegram bot API.
+ */
+public class TelegramConsumer extends ScheduledPollConsumer {
+
+    private TelegramEndpoint endpoint;
+
+    /**
+     * Holds the current offset, used for retrieving incremental updates.
+     */
+    private volatile Long offset;
+
+    public TelegramConsumer(TelegramEndpoint endpoint, Processor processor) {
+        super(endpoint, processor);
+        setDelay(endpoint.getConfiguration().getDelay());
+        this.endpoint = endpoint;
+    }
+
+    @Override
+    protected int poll() throws Exception {
+
+        TelegramConfiguration config = endpoint.getConfiguration();
+
+        Long realOffset = offset != null ? offset : 0L;
+
+        TelegramService service = TelegramServiceProvider.get().getService();
+
+        log.debug("Polling Telegram service to get updates");
+
+        UpdateResult updateResult = service.getUpdates(config.getAuthorizationToken(), offset, config.getLimit(), config.getTimeout());
+        if (updateResult.getUpdates() == null) {
+            // to simplify processing
+            updateResult.setUpdates(Collections.emptyList());
+        }
+
+        if (!updateResult.isOk()) {
+            throw new IllegalStateException("The server was unable to process the request. Response was " + updateResult);
+        }
+
+        List<Update> updates = updateResult.getUpdates();
+
+        if (updates.size() > 0) {
+            log.info("Received " + updates.size() + " updates from Telegram service");
+        } else {
+            log.debug("No updates received from Telegram service");
+        }
+
+        processUpdates(updates);
+
+        // update offset to prevent retrieving the same data again
+        updateOffset(updates);
+
+        return updates.size();
+    }
+
+    private void processUpdates(List<Update> updates) throws Exception {
+        for (Update update : updates) {
+
+            log.debug("Received update from Telegram service: " + update);
+
+            Exchange exchange = endpoint.createExchange();
+
+            if (update.getMessage() != null) {
+                exchange.getIn().setBody(update.getMessage());
+
+                if (update.getMessage().getChat() != null) {
+                    exchange.getIn().setHeader(TelegramConstants.TELEGRAM_CHAT_ID, update.getMessage().getChat().getId());
+                }
+            }
+            getProcessor().process(exchange);
+        }
+    }
+
+    private void updateOffset(List<Update> updates) {
+        OptionalLong ol = updates.stream().mapToLong(Update::getUpdateId).max();
+        if (ol.isPresent()) {
+            this.offset = ol.getAsLong() + 1;
+            log.debug("Next Telegram offset will be " + this.offset);
```

Comment:
Yes, I tend to avoid using placeholders in SLF4J just because varargs have been introduced recently and it is easy to make mistakes at the ERROR level (due to overloading)... But I agree, this case is safe :)